From 1ed358abc30d36e2aac1221df527ec714f9f7b79 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:53 -0500
Subject: [drm] radeon/dpm: set the thermal type properly for special configs

Message-id: <1418852380-13061-43-git-send-email-rclark@redhat.com>
Patchwork-id: 102218
O-Subject: [RHEL7 drm 42/89] drm/radeon/dpm: set the thermal type properly for special configs
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since v3.16.4

commit ee75424b5bb21945a073e084933b6a6015962ad9

Author:     Alex Deucher <alexander.deucher@amd.com>
AuthorDate: Mon Sep 8 02:33:32 2014 -0400
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:19:54 2014 -0500

    drm/radeon/dpm: set the thermal type properly for special configs

    commit ff4377924f7e587c61bcbc704eafecf6c7bd2e00 upstream.

    On systems with special thermal configurations make sure we make
    note of the thermal setup.  This is required for proper firmware
    configuration on these systems.

    Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/radeon/radeon_atombios.c b/drivers/gpu/drm/radeon/radeon_atombios.c
index 5f19e52..4ad00b1 100644
--- a/drivers/gpu/drm/radeon/radeon_atombios.c
+++ b/drivers/gpu/drm/radeon/radeon_atombios.c
@@ -2281,19 +2281,31 @@ static void radeon_atombios_add_pplib_thermal_controller(struct radeon_device *r
      (controller->ucFanParameters &
       ATOM_PP_FANPARAMETERS_NOFAN) ? "without" : "with");
    rdev->pm.int_thermal_type = THERMAL_TYPE_KV;
-  } else if ((controller->ucType ==
-       ATOM_PP_THERMALCONTROLLER_EXTERNAL_GPIO) ||
-      (controller->ucType ==
-       ATOM_PP_THERMALCONTROLLER_ADT7473_WITH_INTERNAL) ||
-      (controller->ucType ==
-       ATOM_PP_THERMALCONTROLLER_EMC2103_WITH_INTERNAL)) {
-   DRM_INFO("Special thermal controller config\n");
+  } else if (controller->ucType ==
+      ATOM_PP_THERMALCONTROLLER_EXTERNAL_GPIO) {
+   DRM_INFO("External GPIO thermal controller %s fan control\n",
+     (controller->ucFanParameters &
+      ATOM_PP_FANPARAMETERS_NOFAN) ? "without" : "with");
+   rdev->pm.int_thermal_type = THERMAL_TYPE_EXTERNAL_GPIO;
+  } else if (controller->ucType ==
+      ATOM_PP_THERMALCONTROLLER_ADT7473_WITH_INTERNAL) {
+   DRM_INFO("ADT7473 with internal thermal controller %s fan control\n",
+     (controller->ucFanParameters &
+      ATOM_PP_FANPARAMETERS_NOFAN) ? "without" : "with");
+   rdev->pm.int_thermal_type = THERMAL_TYPE_ADT7473_WITH_INTERNAL;
+  } else if (controller->ucType ==
+      ATOM_PP_THERMALCONTROLLER_EMC2103_WITH_INTERNAL) {
+   DRM_INFO("EMC2103 with internal thermal controller %s fan control\n",
+     (controller->ucFanParameters &
+      ATOM_PP_FANPARAMETERS_NOFAN) ? "without" : "with");
+   rdev->pm.int_thermal_type = THERMAL_TYPE_EMC2103_WITH_INTERNAL;
   } else if (controller->ucType < ARRAY_SIZE(pp_lib_thermal_controller_names)) {
    DRM_INFO("Possible %s thermal controller at 0x%02x %s fan control\n",
      pp_lib_thermal_controller_names[controller->ucType],
      controller->ucI2cAddress >> 1,
      (controller->ucFanParameters &
       ATOM_PP_FANPARAMETERS_NOFAN) ? "without" : "with");
+   rdev->pm.int_thermal_type = THERMAL_TYPE_EXTERNAL;
    i2c_bus = radeon_lookup_i2c_gpio(rdev, controller->ucI2cLine);
    rdev->pm.i2c_bus = radeon_i2c_lookup(rdev, &i2c_bus);
    if (rdev->pm.i2c_bus) {
-- 
1.7.1